core/vcpu.c

```diff
@@ -2622,10 +2619,31 @@ static void handle_cpuid_virtual(struct vcpu_t *vcpu, uint32_t a, uint32_t c)
              * that is an old i7 system, so the emulator can still utilize the
              * enough extended features of the hardware, but doesn't crash.
              */
-            hw_family = ((state->_eax >> 16) & 0xFF0) |
-                        ((state->_eax >> 8) & 0xF);
-            hw_model = ((state->_eax >> 12) & 0xF0) |
-                       ((state->_eax >> 4) & 0xF);
+            union cpuid_1_eax {
+                uint32_t raw;
+                struct {
+                    uint32_t steppingID    : 4;
+                    uint32_t model         : 4;
+                    uint32_t familyID      : 4;
+                    uint32_t processorType : 2;
+                    uint32_t reserved      : 2;
+                    uint32_t extModelID    : 4;
+                    uint32_t extFamilyID   : 8;
+                    uint32_t reserved2     : 4;
+                };
+            } cpuid_eax;
+            cpuid_eax.raw = state->_eax;
+
+            if (0xF != cpuid_eax.familyID)
+                hw_family = cpuid_eax.familyID;
+            else
+                hw_family = cpuid_eax.familyID + (cpuid_eax.extFamilyID << 4);
+
+            if (0x6 == cpuid_eax.familyID || 0xF == cpuid_eax.familyID)
+                hw_model = (cpuid_eax.extModelID << 4) + cpuid_eax.model;
+            else
+                hw_model = cpuid_eax.model;
+
```

Comment:
The refactoring makes it much clean now and match the pseudo code of SDM better.